coreax/kernels/base.py

```diff
@@ -0,0 +1,503 @@
+:math:`\frac{1}{\sqrt{2\pi} \,*\, \text{length_scale}}`, yields the well known
+Gaussian kernel.
+
+A :class:`~coreax.kernels.ScalarValuedKernel` must implement a
+:meth:`~coreax.kernels.ScalarValuedKernel.compute_elementwise` method, which returns the
+floating point value after evaluating the kernel on two floats, ``x`` and ``y``.
+Additional methods, such as :meth:`ScalarValuedKernel.grad_x_elementwise`, can
+optionally be overridden to improve performance. The canonical example is when a
+suitable closed-form representation of a higher-order gradient can be used to avoid the
+expense of performing automatic differentiation.
+
+Such an example can be seen in :class:`coreax.kernels.SteinKernel`, where the analytic
+forms of :meth:`ScalarValuedKernel.divergence_x_grad_y` are significantly cheaper to
+compute than the automatic differentiated default.
+"""
+
+# Support class typing annotations inside itself
+from __future__ import annotations
+
+from abc import abstractmethod
+from typing import Union
+
+import equinox as eqx
+import jax
+import jax.numpy as jnp
+import jax.tree_util as jtu
+from jax import Array, grad, jacrev
+from jax.typing import ArrayLike
+from typing_extensions import override
+
+from coreax.data import Data, is_data
+from coreax.kernels.util import block_data_convert
+from coreax.util import pairwise, tree_leaves_repeat
+
+
+class ScalarValuedKernel(eqx.Module):
+    """Abstract base class for scalar-valued kernels."""
+
+    def __add__(self, addition: ScalarValuedKernel) -> AdditiveKernel:
+        """Overload `+` operator."""
+        if not isinstance(addition, ScalarValuedKernel):
+            raise ValueError("'addition' must be an instance of a 'ScalarValuedKernel'")
+        return AdditiveKernel(self, addition)
+
+    def __mul__(self, product: ScalarValuedKernel) -> ProductKernel:
+        """Overload `*` operator."""
+        if not isinstance(product, ScalarValuedKernel):
+            raise ValueError("'product' must be an instance of a 'ScalarValuedKernel'")
+        return ProductKernel(self, product)
+
+    def __pow__(self, power: int) -> ProductKernel:
+        """Overload `**` operator."""
+        min_power = 2
+        power = int(power)
+        if power < min_power:
+            raise ValueError("'power' must be an integer greater than or equal to 2.")
+
+        first_kernel = self
+        second_kernel = self
+
+        # Ensure the first and second kernels are symmetric for even powers to make use
+        # of reduced computation capabilities in ProductKernel. For example,
+        # :meth:`~divergence_x_grad_y_elementwise` can be computed more efficiently if
+        # the first and second kernels are recognised as the same.
+        for i in range(min_power, power):
+            if i % 2 == 0:
+                first_kernel = ProductKernel(first_kernel, self)
+            else:
+                second_kernel = ProductKernel(second_kernel, self)
```

Comment:
Would it be more efficient to instead implement a `PowerKernel` subclassing `UniCompositeKernel` that computes this directly, rather than relying on nested `ProductKernel` instances? That would also allow non-integer powers. That's probably out-of-scope for this PR, though, so may want to be its own issue.

Reply:
Yes that seems like a good idea, will try implement this

Reply:
Implemented this now